Strict repository lock handling

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
restic operations did not check that their lock is still valid. This can lead to a situation where for example a backup client is paused, then in the meantime `unlock` is called for the repository, which removes the now stale lock, and some time later the backup continues. If in the meantime `prune` was called, then this results in a broken snapshot.

This PR changes the locking behavior to actually be strict. If restic is unable to refresh locks in time, then the whole operation will be cancelled. This is done by tying the context used by restic's commands to the lock lifetime. If the monitoring goroutine for the lock detects that the lock file was not refresh in time, then the context will be canceled. Thereby, the command is forcibly terminated.

To keep the implementation simple there are now two goroutines per lock. One which periodically refreshes the lock file and one which monitors that the expiry is done in time. The time limit to refresh the lock file is a few minutes shorter than the duration after which a lock file becomes stale. This is intended to compensate for a small amount of clock drift between clients.

Most code changes revolve around cleaning up the usage of the global context previously available via `globalOptions.ctx`. Depending on the command either that context was accessed via `gopts.ctx` or a local copy in `ctx`. Alternatively, a few commands also introduce an additional context using `context.WithCancel`. The latter is unnecessary as the global context is canceled when restic shuts down. In total, these different ways to use the context made it non-obvious which contexts have to be tied to the lock lifetime to ensure that the command properly terminates after a failed lock refresh.

This PR solves the "additional behavior changes" proposed in #2715. However, it does not check whether a lock file disappeared or not as the could lead to race conditions with storage backends which do not provide strongly consistent directory listings.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Fixes #2715

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

MichaelEischer committed:
Remove unnecessary context.WithCancel calls

The gopts.ctx is cancelled when the main() method of restic exits.

MichaelEischer committed:
key: Cleanup method signatures

MichaelEischer committed:
Remove ctx from globalOptions

Previously the global context was either accessed via gopts.ctx,
stored in a local variable and then used within that function or
sometimes both. This makes it very hard to follow which ctx or a wrapped
version of it reaches which method.

Thus just drop the context from the globalOptions struct and pass it
explicitly to every command line handler method.

MichaelEischer committed:
Prepare for context bound to lock lifetime

MichaelEischer committed:
lock: Implement strict lock expiry monitoring

Restic continued e.g. a backup task even when it failed to renew the
lock or failed to do so in time. For example if a backup client enters
standby during the backup this can allow other operations like `prune`
to run in the meantime (after calling `unlock`). After leaving standby
the backup client will continue its backup and upload indexes which
refer pack files that were removed in the meantime.

This commit introduces a goroutine explicitly monitoring for locks that
are not refreshed in time. To simplify the implementation there's now a
separate goroutine to refresh the lock and monitor for timeouts for each
lock. The monitoring goroutine would now cause the backup to fail as the
client has lost it's lock in the meantime.

The lock refresh goroutines are bound to the context used to lock the
repository initially. The context returned by `lockRepo` is also
cancelled when any of the goroutines exits. This ensures that the
context is cancelled whenever for any reason the lock is no longer
refreshed.

MichaelEischer committed:
lock: Use repository interface instead of struct

MichaelEischer committed:
lock: Add integration test

The tests check that the wrapped context is properly canceled whenever
the repository is unlock or when the lock refresh fails.

MichaelEischer committed:
add changelog

MichaelEischer committed:
lock: Do not ignore invalid lock files

While searching for lock file from concurrently running restic
instances, restic ignored unreadable lock files. These can either be
in fact invalid or just be temporarily unreadable. As it is not really
possible to differentiate between both cases, just err on the side of
caution and consider the repository as already locked.

The code retries searching for other locks up to three times to smooth
out temporarily unreadable lock files.

MichaelEischer commented:
I've added an extra commit which lets restic fail if it is unable to read any of the lock files in the repository. The idea is to prevent overlooking concurrent restic processes if a lock file is unreadable for some reason.

XLTechie commented:
+1! This seems like a critical change.
I have recently started using restic on multiple large servers, backing up to a cloud storage provider on a schedule, and this almost immediately became a concern.
I hope this can be resolved and merged soon!

MichaelEischer commented:
> I have recently started using restic on multiple large servers, backing up to a cloud storage provider on a schedule, and this almost immediately became a concern.

In what regard? For servers, which don't hibernate/use standby, this PR is likely not too relevant.

MichaelEischer committed:
lock: fix timer expiry monitoring during standby

Monotonic timers are paused during standby. Thus these timers won't fire
after waking up. Fall back to periodic polling to detect too large clock
jumps. See https://github.com/golang/go/issues/35012 for a discussion of
go timers during standby.

MichaelEischer commented:
I've changed the timer expiry checks as apparently timers are stopped during standby: https://github.com/golang/go/issues/35012

XLTechie commented:
I should have said "virtual machine" rather than "server".

In one case (the reason I started using Restic, in fact), the cloud provider 
has started becoming unstable. (I.E. the company is failing as a concern and its 
support has become unreachable).

VMs can restart just fine, and a few minutes of downtime here and there 
isn't critical for the functions they serve, but loss of data is.

Because of the prepaid contracts we have with them, I am loathed to just dump 
and move to a more reliable provider such as Linode, until closer to the paid 
expiration, early next year.

But the potential for a backup to be interrupted in that context is very strong.

Unlike, say, on my home machine, where I also use it. If that fails out for some 
reason, I will absolutely know about it and can take manual corrective action if 
necessary.

fd0 reviewed (approved):
I haven't done an in-depth review of the new lock code, but looked at the design and I like it a lot!

MichaelEischer committed:
pass global context through cobra

MichaelEischer commented:
The globalContext is now passed through cobra. (credit goes to @fd0)

virtadpt commented:
> I should have said "virtual machine" rather than "server".

It's not just virtual machines.  I started seeing this on my physical servers a couple of days ago.

MichaelEischer commented:
Please have a look at #4199 or #4262
